File: src/praisonai-agents/praisonaiagents/model_harness/profiles.py
"""Harness profiles and the model-family resolver.

A :class:`HarnessProfile` bundles two family-tuned knobs:

* ``base_prompt`` — an optional system-prompt fragment prepended to the
  assembled prompt with family-specific tool-use guidance.
* ``preferred_edit_format`` — the file-edit primitive the family handles
  best (``"apply_patch"`` or ``"edit_file"``); ``None`` keeps the current
  both-exposed behaviour.

The registry maps case-insensitive substring matchers (evaluated against the
model id) to profiles. Unknown models fall back to :data:`DEFAULT_PROFILE`,
which is behaviour-preserving (no prompt fragment, no edit-format preference).
"""

from __future__ import annotations

import threading
from dataclasses import dataclass
from typing import List, Optional, Tuple

try:  # Python 3.8+ Protocol lives in typing; fall back for older interpreters.
    from typing import Protocol, runtime_checkable
except ImportError:  # pragma: no cover - defensive
    from typing_extensions import Protocol, runtime_checkable  # type: ignore


@dataclass(frozen=True)
class HarnessProfile:
    """A model-family harness profile.

    Attributes:
        name: Identifier for the profile (e.g. ``"default"``, ``"anthropic"``).
        base_prompt: Optional prompt fragment prepended to the system prompt.
            ``None`` means no fragment (behaviour-preserving default).
        preferred_edit_format: Preferred edit primitive name, or ``None`` to
            keep exposing both primitives in their current order.
    """

    name: str = "default"
    base_prompt: Optional[str] = None
    preferred_edit_format: Optional[str] = None


@runtime_checkable
class HarnessResolverProtocol(Protocol):
    """Protocol for objects that resolve a model id to a harness profile."""

    def resolve_harness(self, model: Optional[str]) -> HarnessProfile:
        ...


# Behaviour-preserving default: no fragment, no edit-format preference.
DEFAULT_PROFILE = HarnessProfile(name="default")


# Family matchers: (list of case-insensitive substrings, profile).
# Ordered; first match wins. Kept intentionally small and data-driven so
# consumers can extend/override via register_profile().
_DEFAULT_REGISTRY: List[Tuple[List[str], HarnessProfile]] = [
    (
        ["claude", "anthropic"],
        HarnessProfile(
            name="anthropic",
            base_prompt=(
                "When editing files, prefer patch-style edits: use apply_patch "
                "to create or rewrite files and edit_file for targeted changes."
            ),
            preferred_edit_format="apply_patch",
        ),
    ),
    (
        ["gpt", "openai", "o1", "o3", "o4"],
        HarnessProfile(
            name="openai",
            base_prompt=(
                "When editing files, prefer targeted string-replacement edits: "
                "use edit_file to modify existing files precisely."
            ),
            preferred_edit_format="edit_file",
        ),
    ),
]

_registry_lock = threading.RLock()
_registry: List[Tuple[List[str], HarnessProfile]] = list(_DEFAULT_REGISTRY)


def register_profile(matchers: List[str], profile: HarnessProfile) -> None:
    """Register (or prepend) a family matcher → profile mapping.

    New registrations take precedence over the built-in defaults so callers
    can override behaviour. Thread-safe.

    Args:
        matchers: Case-insensitive substrings matched against the model id.
        profile: The :class:`HarnessProfile` to resolve when a matcher hits.
    """
    normalized = [m.lower() for m in matchers if m]
    with _registry_lock:
        _registry.insert(0, (normalized, profile))


def resolve_harness(model: Optional[str]) -> HarnessProfile:
    """Resolve a model id to a :class:`HarnessProfile`.

    Unknown / falsy models return :data:`DEFAULT_PROFILE` (no behaviour change).

    Args:
        model: The model id string (e.g. ``"claude-opus-4"``).

    Returns:
        The first matching profile, or the default profile.
    """
    if not model:
        return DEFAULT_PROFILE
    lowered = model.lower()
    with _registry_lock:
        for matchers, profile in _registry:
            if any(m in lowered for m in matchers):
                return profile
    return DEFAULT_PROFILE

File: src/praisonai-ts/src/model-harness/profiles.ts
/**
 * Harness profiles and the model-family resolver.
 *
 * Port of `praisonaiagents/model_harness/profiles.py`.
 *
 * A {@link HarnessProfile} bundles two family-tuned knobs:
 *
 * - `basePrompt` — an optional system-prompt fragment prepended to the
 *   assembled prompt with family-specific tool-use guidance.
 * - `preferredEditFormat` — the file-edit primitive the family handles best
 *   (`"apply_patch"` or `"edit_file"`); `null` keeps the both-exposed
 *   behaviour.
 *
 * The registry maps case-insensitive substring matchers (evaluated against
 * the model id) to profiles. Unknown models fall back to
 * {@link DEFAULT_PROFILE}, which is behaviour-preserving (no prompt
 * fragment, no edit-format preference).
 *
 * Python's `threading.RLock` is omitted: the registry is a plain array on a
 * single-threaded runtime.
 */

/** Constructor options for {@link HarnessProfile} (Python dataclass fields). */
export interface HarnessProfileConfig {
  /** Identifier for the profile (e.g. `"default"`, `"anthropic"`). Default `"default"`. */
  name?: string;
  /** Optional prompt fragment prepended to the system prompt. `null` means none. */
  basePrompt?: string | null;
  /** Preferred edit primitive name, or `null` to keep exposing both primitives. */
  preferredEditFormat?: string | null;
}

/**
 * A model-family harness profile.
 *
 * Python parity: `model_harness/profiles.py::HarnessProfile` (frozen
 * dataclass). Instances are frozen after construction, matching
 * `@dataclass(frozen=True)`.
 */
export class HarnessProfile {
  readonly name: string;
  readonly basePrompt: string | null;
  readonly preferredEditFormat: string | null;

  constructor(config: HarnessProfileConfig = {}) {
    this.name = config.name ?? 'default';
    this.basePrompt = config.basePrompt ?? null;
    this.preferredEditFormat = config.preferredEditFormat ?? null;
    Object.freeze(this);
  }
}

/**
 * Protocol for objects that resolve a model id to a harness profile.
 *
 * Python parity: `model_harness/profiles.py::HarnessResolverProtocol`.
 */
export interface HarnessResolverProtocol {
  resolveHarness(model?: string | null): HarnessProfile;
}

/** Behaviour-preserving default: no fragment, no edit-format preference. */
export const DEFAULT_PROFILE: HarnessProfile = new HarnessProfile({ name: 'default' });

/** A registry entry: `[case-insensitive substrings, profile]`. First match wins. */
export type HarnessRegistryEntry = readonly [ReadonlyArray<string>, HarnessProfile];

/**
 * Family matchers. Ordered; first match wins. Kept intentionally small and
 * data-driven so consumers can extend/override via {@link registerProfile}.
 */
const _DEFAULT_REGISTRY: ReadonlyArray<HarnessRegistryEntry> = [
  [
    ['claude', 'anthropic'],
    new HarnessProfile({
      name: 'anthropic',
      basePrompt:
        'When editing files, prefer patch-style edits: use apply_patch ' +
        'to create or rewrite files and edit_file for targeted changes.',
      preferredEditFormat: 'apply_patch',
    }),
  ],
  [
    ['gpt', 'openai', 'o1', 'o3', 'o4'],
    new HarnessProfile({
      name: 'openai',
      basePrompt:
        'When editing files, prefer targeted string-replacement edits: ' +
        'use edit_file to modify existing files precisely.',
      preferredEditFormat: 'edit_file',
    }),
  ],
];

let _registry: HarnessRegistryEntry[] = [..._DEFAULT_REGISTRY];

/**
 * Register (prepend) a family matcher → profile mapping.
 *
 * New registrations take precedence over the built-in defaults so callers
 * can override behaviour.
 *
 * Python parity: `model_harness/profiles.py::register_profile`.
 *
 * @param matchers Case-insensitive substrings matched against the model id.
 *   Empty strings are dropped, as in Python; an entry with no matchers never
 *   matches.
 * @param profile The profile to resolve when a matcher hits.
 */
export function registerProfile(matchers: string[], profile: HarnessProfile): void {
  const normalized = matchers.filter((m) => Boolean(m)).map((m) => m.toLowerCase());
  _registry.unshift([normalized, profile]);
}

/**
 * Resolve a model id to a {@link HarnessProfile}.
 *
 * Unknown / falsy models return {@link DEFAULT_PROFILE} (no behaviour change).
 *
 * Python parity: `model_harness/profiles.py::resolve_harness`.
 */
export function resolveHarness(model?: string | null): HarnessProfile {
  if (!model) return DEFAULT_PROFILE;
  const lowered = model.toLowerCase();
  for (const [matchers, profile] of _registry) {
    if (matchers.some((m) => lowered.includes(m))) {
      return profile;
    }
  }
  return DEFAULT_PROFILE;
}

/** Snapshot of the current registry, most-recently-registered first (TypeScript-only). */
export function listHarnessProfiles(): ReadonlyArray<HarnessRegistryEntry> {
  return _registry.map(([matchers, profile]) => [[...matchers], profile] as const);
}

/**
 * Restore the built-in registry, discarding every {@link registerProfile}
 * call. TypeScript-only test helper; Python has no equivalent (its tests
 * patch the module-level list directly).
 */
export function resetHarnessRegistry(): void {
  _registry = [..._DEFAULT_REGISTRY];
}
